### PR TITLE
Add docs for GCS sources and sinks

### PR DESCRIPTION
**Pull request by albertshau**
Added documentation for GCS sources and sinks. Also performed
a bunch of cleanup to remove bad properties and redundant code.

Also fixed the GCS parquet file sink that was completely broken.

### CODE REVIEW

File: docs/GCSAvro-batchsink.md [hunk 1]
```diff
@@ -1,3 +1,44 @@
...
+
+**Base Path:** The path to write to. For example, gs://<bucket>/path/to/directory
+
+**Path Suffix:** The time format for the output directory that will be appended to the path.
+For example, the format 'yyyy-MM-dd-HH-mm' will result in a directory of the form '2015-01-01-20-42'.
+If not specified, nothing will be appended to the path."
+
+**Schema:** The schema of records to write.
+
+**Codec:** The codec to use when writing data. Must be 'none', 'snappy', 'gzip', or 'lzo'.
```
Comment (gsps1) [inline docs/GCSAvro-batchsink.md]: will be good to mention what is the default codec

File: docs/GCSFile-batchsource.md [hunk 2]
```diff
@@ -1,6 +1,56 @@
...
+
+**Path:** Path to file(s) to be read. If a directory is specified, terminate the path name with a '/'.
+For example, gs://<bucket>/path/to/directory/
+
+**Regex Path Filter:** Regex to filter out files in the path. It accepts regular expression which is applied to the complete
+path and returns the list of files that match the specified pattern.
+
+**Path Field:** If specified, each output record will include a field with this name that contains the file URI
+that the record was read from. Requires a customized version of CombineFileInputFormat, so it cannot be used if
+an inputFormatClass is given.
```
Comment (gsps1) [inline docs/GCSFile-batchsource.md]: where is the inputFormatClass given ? i do not see that property in the config.

File: docs/GCSParquet-batchsink.md [hunk 3]
```diff
@@ -1,3 +1,44 @@
...
+
+If the plugin is not run on a Dataproc cluster, the path to a service account key must be provided.
+The service account key can be found on the Dashboard in the Cloud Platform Console.
+Make sure the account key has permission to access BigQuery and Google Cloud Storage.
+The service account key file needs to be available on every node in your cluster and
+must be readable by all users running the job.
+
+Properties
+----------
+**Reference Name:** This will be used to uniquely identify this source for lineage, annotating metadata, etc.
```
Comment (gsps1) [inline docs/GCSParquet-batchsink.md]: this `sink`

File: docs/GCSParquet-batchsink.md [hunk 4]
```diff
@@ -1,3 +1,44 @@
...
+
+**Project ID**: The Google Cloud Project ID, which uniquely identifies a project.
+It can be found on the Dashboard in the Google Cloud Platform Console.
+
+**Service Account File Path**: Path on the local file system of the service account key used for
+authorization. Does not need to be specified when running on a Dataproc cluster.
+When running on other clusters, the file must be present on every node in the cluster.
+
+**Bucket Name**: The bucket to read from.
```
Comment (gsps1) [inline docs/GCSParquet-batchsink.md]: write

File: docs/GCSText-batchsink.md [hunk 5]
```diff
@@ -1,4 +1,42 @@
...
+
+If the plugin is not run on a Dataproc cluster, the path to a service account key must be provided.
+The service account key can be found on the Dashboard in the Cloud Platform Console.
+Make sure the account key has permission to access BigQuery and Google Cloud Storage.
+The service account key file needs to be available on every node in your cluster and
+must be readable by all users running the job.
+
+Properties
+----------
+**Reference Name:** This will be used to uniquely identify this source for lineage, annotating metadata, etc.
```
Comment (gsps1) [inline docs/GCSText-batchsink.md]: sink

File: docs/GCSText-batchsink.md [hunk 6]
```diff
@@ -1,4 +1,42 @@
...
+
+**Project ID**: The Google Cloud Project ID, which uniquely identifies a project.
+It can be found on the Dashboard in the Google Cloud Platform Console.
+
+**Service Account File Path**: Path on the local file system of the service account key used for
+authorization. Does not need to be specified when running on a Dataproc cluster.
+When running on other clusters, the file must be present on every node in the cluster.
+
+**Bucket Name**: The bucket to read from.
```
Comment (gsps1) [inline docs/GCSText-batchsink.md]: write to

File: docs/GCSText-batchsink.md [hunk 7]
```diff
@@ -1,4 +1,42 @@
...
+
+**Base Path:** The path to write to. For example, gs://<bucket>/path/to/directory
+
+**Path Suffix:** The time format for the output directory that will be appended to the path.
+For example, the format 'yyyy-MM-dd-HH-mm' will result in a directory of the form '2015-01-01-20-42'.
+If not specified, nothing will be appended to the path."
+
+**File Type:** The type of file to write.
```
Comment (gsps1) [inline docs/GCSText-batchsink.md]: right now it supports 5 types, will be good to specify them.

File: src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java [hunk 8]
```diff
@@ -29,29 +29,23 @@
 import co.cask.cdap.etl.api.batch.BatchSinkContext;
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java]: there is still couple of unused import in this class

File: src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java [hunk 9]
```diff
@@ -29,29 +29,23 @@
...
 import java.util.Map;
 import javax.annotation.Nullable;
 
 /**
- * {@link GCSAvroBatchSink} that stores data in avro format to S3.
+ * {@link GCSAvroBatchSink} that stores data in avro format to Google Cloud Storage.
  */
 @Plugin(type = BatchSink.PLUGIN_TYPE)
 @Name("GCSAvro")
 @Description("GCS Avro Batch Sink.")
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java]: can add more description ?

File: src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java [hunk 10]
```diff
@@ -66,77 +60,37 @@ public void initialize(BatchRuntimeContext context) throws Exception {
...
   public static class GCSAvroSinkConfig extends GCSBatchSinkConfig {
     @Name("schema")
-    @Description("Specifies the schema to used for defining the GCS Dataset.")
+    @Description("The schema of records to write.")
     @Macro
     protected String schema;
 
     @Name("codec")
-    @Description("Specifies compression codec to be used.")
+    @Description("The compression codec to use when writing data. Must be 'none', 'snappy', 'gzip', or 'lzo'.")
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java]: looked at the code and the UI widget, it supports only "none", "snappy" and "deflate"

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java [hunk 11]
```diff
@@ -80,103 +61,79 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+    outputConfig.put(FileOutputFormat.OUTDIR, config.getOutputDir(context.getLogicalStartTime()));
+    if (config.serviceFilePath != null) {
+      outputConfig.put("mapred.bq.auth.service.account.json.keyfile", config.serviceFilePath);
+      outputConfig.put("google.cloud.auth.service.account.json.keyfile", config.serviceFilePath);
     }
-    context.addOutput(Output.of(config.referenceName, new SinkOutputFormatProvider(
-      outputFormatProvider.getOutputFormatClassName(), outputConfig)));
+    outputConfig.put("fs.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem");
+    outputConfig.put("fs.AbstractFileSystem.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS");
+    String projectId = config.project == null ? ServiceOptions.getDefaultProjectId() : config.project;
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: better to call `GCPUtil.getProjectId(..)` as that will throw exception if project id cannot be detected from environment.

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java [hunk 12]
```diff
@@ -80,103 +61,79 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
-      this.project = project;
-      this.serviceAccountFilePath = serviceAccountFilePath;
-      this.bucket = bucket;
-      this.properties = updateFileSystemProperties(path, properties);
     }
 
     public void validate() {
+      // validate project id availability
+      GCPUtil.getProjectId(project);
       if (path != null && !path.startsWith("gs://")) {
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: should check projectId, path are not a macro.

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java [hunk 13]
```diff
@@ -80,103 +61,79 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
         throw new IllegalArgumentException("Path must start with gs://.");
       }
     }
+
+    protected String getOutputDir(long logicalStartTime) {
+      String timeSuffix = !Strings.isNullOrEmpty(suffix) ? new SimpleDateFormat(suffix).format(logicalStartTime) : "";
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: will be good to check `suffix` is valid during `validate`

File: src/main/java/co/cask/gcs/sink/GCSParquetBatchSink.java [hunk 14]
```diff
@@ -61,68 +57,35 @@ public void initialize(BatchRuntimeContext context) throws Exception {
...
   public static class GCSParquetSinkConfig extends GCSBatchSinkConfig {
     @Name("schema")
-    @Description("Specifies the schema to used for defining the GCS Dataset.")
+    @Description("The schema of records to write.")
     @Macro
     protected String schema;
 
     @Nullable
-    @Description("Used to specify the compression codec to be used for the final dataset.")
+    @Description("The compression codec to use when writing data. Must be 'none', 'snappy', 'gzip', or 'lzo'.")
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSParquetBatchSink.java:88]: similar comment about supported compression types.

File: src/main/java/co/cask/gcs/sink/GCSTextBatchSink.java [hunk 15]
```diff
@@ -88,27 +77,37 @@ public void transform(StructuredRecord input,
     }
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSTextBatchSink.java]: not related to your change, but not sure if its desirable to silently consume the Exception.

File: src/main/java/co/cask/gcs/source/GCSFileBlob.java [hunk 16]
```diff
@@ -50,7 +52,7 @@
  */
 @Plugin(type = BatchSource.PLUGIN_TYPE)
 @Name("GCSFileBlob")
-@Description("Reads content of the whole file as one record")
+@Description("reads the entire content of a Google Cloud Storage object into a single record.")
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/source/GCSFileBlob.java]: `Reads ..` caps

**Review by gsps1 — COMMENTED** [review]
few comments

File: docs/GCSAvro-batchsink.md (hunk 1 above)
Comment (albertshau) [reply docs/GCSAvro-batchsink.md]: done

File: docs/GCSFile-batchsource.md (hunk 2 above)
Comment (albertshau) [reply docs/GCSFile-batchsource.md]: good catch, this was leftover from copy-paste. Removed.

File: docs/GCSParquet-batchsink.md (hunk 3 above)
Comment (albertshau) [reply docs/GCSParquet-batchsink.md]: done

File: docs/GCSParquet-batchsink.md (hunk 4 above)
Comment (albertshau) [reply docs/GCSParquet-batchsink.md]: done

File: docs/GCSText-batchsink.md (hunk 5 above)
Comment (albertshau) [reply docs/GCSText-batchsink.md]: done

File: docs/GCSText-batchsink.md (hunk 6 above)
Comment (albertshau) [reply docs/GCSText-batchsink.md]: done

File: src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java (hunk 8 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java]: done

File: docs/GCSText-batchsink.md (hunk 7 above)
Comment (albertshau) [reply docs/GCSText-batchsink.md]: done

File: src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java (hunk 9 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java]: done

File: src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java (hunk 10 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSAvroBatchSink.java]: done

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java (hunk 11 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: done

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java (hunk 12 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: done

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java (hunk 13 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: done

File: src/main/java/co/cask/gcs/sink/GCSParquetBatchSink.java (hunk 14 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSParquetBatchSink.java:88]: this is correct for parquet. avro and parquet support different codecs apparently

File: src/main/java/co/cask/gcs/sink/GCSTextBatchSink.java (hunk 15 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSTextBatchSink.java]: I agree, removing this.

File: src/main/java/co/cask/gcs/source/GCSFileBlob.java (hunk 16 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/source/GCSFileBlob.java]: done

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java [hunk 17]
```diff
@@ -125,10 +125,15 @@ public GCSBatchSinkConfig() {
 
     public void validate() {
       // validate project id availability
-      GCPUtil.getProjectId(project);
-      if (path != null && !path.startsWith("gs://")) {
+      if (!containsMacro("project")) {
+        GCPUtil.getProjectId(project);
+      }
```
Comment (gsps1) [inline src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: just realized, for other plugins we decided to validate projectId only during `prepareRun` and not during `config` as users could prefer to not provide project for CDAP local sandbox but pipelines configured to run on data-proc where its picked up automatically.

**Review by gsps1 — APPROVED** [review]
just one comment, LGTM otherwise.

File: src/main/java/co/cask/gcs/sink/GCSBatchSink.java (hunk 17 above)
Comment (albertshau) [reply src/main/java/co/cask/gcs/sink/GCSBatchSink.java]: good catch, will remove this check

**Pull request merged**